cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright © 2020 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.proto;
+
+import javax.annotation.Nullable;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Represent preferences
+ */
+public class PreferencesDetail {
+  private final Map<String, String> properties;
+  /**
+   * Sequence id of operations on the preferences
+   */
+  private final Long seqId;
+  /**
+   * Whether it is a resolved preferences or not.
+   */
+  private boolean resolved;
+
+  public static PreferencesDetail merge(PreferencesDetail left, PreferencesDetail right) {
+    if (left == null && right == null) {
+      return null;
+    } else if (left == null) {
+      return right;
+    } else if (right == null) {
+      return left;
+    }
+
+    Map<String, String> properties = new HashMap<>();
+    properties.putAll(left.getProperties());
+    properties.putAll(right.getProperties());
+
+    Long seqId = null;
+    Long leftSeqId = left.getSeqId();
+    Long rightSeqId = right.getSeqId();
+    if (leftSeqId != null && rightSeqId != null) {
+      seqId = Long.max(leftSeqId.longValue(), rightSeqId.longValue());
+    } else if (leftSeqId != null) {
+      seqId = new Long(leftSeqId.longValue());
+    } else if (rightSeqId != null) {
+      seqId = new Long(rightSeqId.longValue());
+    }
+
+    boolean resolved = (left.resolved || right.resolved);
+    return new PreferencesDetail(properties, seqId, resolved);
+  }
+
+  public PreferencesDetail(Map<String, String> properties, @Nullable Long seqId, boolean resolved) {
```

Comment:
I don't think we should allow seqId to be null, as that isn't a valid representation of stored preferences. It forces the caller of getSeqId() to handle a null, which shouldn't ever be the case

Reply:
Done

Reply:
Actually seqId could be null when preferences never get set on the entity. Given the fact that seqId > 0, I am using 0 as default value when preferences are not present. This is consistent with current contact that property map is empty when preferences are not set.

Reply:
I see, makes sense

Reply:
I see. What if preferences are set and then deleted?

Reply:
Deletion just clear the map but still bump up the seqId. So the only case seqId is not there is the entity never had any preferences set on it.